src/Microsoft.AspNetCore.Server.Kestrel/Http/FrameOfT.cs

```diff
@@ -37,8 +37,14 @@ public class Frame<TContext> : Frame
                     {
                         if (SocketInput.RemoteIntakeFin)
                         {
+                            if (TakeStartLine(SocketInput))
```

Comment:
It does scan again. Since `TakeStartLine` was last called before checking whether RemoteIntakeFin is true, there might be more data available.

I fixed the two places that affected the failing test referenced in the issue, but there are actually a few other places the race still exists that I intend to fix. It would nice to have a better pattern for consuming input that made this race impossible. Maybe it's as simple as checking if RemoteIntakeFin is true first and then trying to consume input.